src/Common/src/System/Runtime/InteropServices/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Allows limited thread safe reuse of heap buffers to limit memory pressure.
+    /// 
+    /// This cache does not ensure that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
+    {
+        private readonly ulong _minSize;
+        private readonly ulong _maxSize;
+
+        private SafeHeapHandle[] _handleCache;
+
+        /// <param name="minSize">Smallest buffer size to allocate in bytes.</param>
+        /// <param name="maxSize">The largest buffer size to cache in bytes.</param>
+        /// <param name="maxHandles">The maximum number of handles to cache.</param>
+        public SafeHeapHandleCache(ulong minSize = 64, ulong maxSize = 1024 * 2, int maxHandles = 0)
+        {
+            _minSize = minSize;
+            _maxSize = maxSize;
+            _handleCache = new SafeHeapHandle[maxHandles > 0 ? maxHandles : Environment.ProcessorCount * 4];
+        }
+
+        /// <summary>
+        /// Get a HeapHandle
+        /// </summary>
+        public SafeHeapHandle Acquire(ulong minSize = 0)
+        {
+            if (minSize < _minSize) minSize = _minSize;
+
+            SafeHeapHandle handle = null;
+
+            for (int i = 0; i < _handleCache.Length; i++)
+            {
+                handle = Interlocked.Exchange(ref _handleCache[i], null);
+                if (handle != null) break;
+            }
+
+            if (handle != null)
+            {
+                if (handle.ByteLength < minSize)
+                    handle.Resize(minSize);
```

Comment:
We could end up in a situation where there's a valid handle of a fine size at index i+1, but we grab the one at index i and need to resize it.  Something to think about for later.